py3status/modules/random_wallpaper.py

```diff
@@ -0,0 +1,302 @@
+        self.__wallpapers = self.__find_wallpapers()
+        if self.first_image_path is None:
+            # get random index
+            self.__current_index = randint(0, len(self.__wallpapers) - 1)
+        else:
+            # get index of given path
+            try:
+                self.__current_index = \
+                    self.__wallpapers.index(self.first_image_path)
+            except ValueError:
+                self.__log_warning("Tried to force first image but path %s was"
+                                   "not found in the list of wallpapers"
+                                   % self.first_image_path)
+                self.__current_index = 0
+        self.__set_wallpaper(self.__wallpapers[self.__current_index])
+
+    # Private Methods
+
+    def __init__(self):
+        # __feh = path to the feh executable
+        self.__feh = None
+        # __wallpapers = list of found (and suitable) wallpapers
+        self.__wallpapers = None
+        # __current_name = path to the currently loaded wallpaper
+        self.__current_name = None
+        # __current_index = index of the current wallpaper (in __wallpapers)
+        self.__current_index = None
+
+    def __find_feh(self):
+        """
+        Returns the path to the feh executable, or None if feh is not found.
+        """
+        process = subprocess.Popen(['which', 'feh'], stdout=subprocess.PIPE,
+                                   stderr=subprocess.PIPE)
+        process.wait()
+        code = process.returncode
+        if code != 0:
+            err = process.stderr.read()
+            self.__log_error("Could not find feh: %d %s" % (code, err))
+            return None  # Error
+        else:
+            result = process.stdout.read()[:-1]
+            # The last character is '\n'
+            return result
+
+    def __find_wallpapers(self):
+        """
+        Returns an array containing the paths to the wallpapers found in the
+        specified directories.
+        """
+        wallpapers = list()
+        for dirpath in self.search_dirs:
+            self.__find_wallpapers_in_dir(wallpapers, dirpath)
+        return wallpapers
+
+    def __find_wallpapers_in_dir(self, wallpapers, dirpath):
+        """
+        Adds the files in the given directory to wallpapers.
+        wallpapers = an initialized list (potentially modified)
+        dirpath = Path to the directory that will be explored. This path will
+        be expanded using expanduser so '~' is authorized.
+        """
+        dirpath = os.path.expanduser(dirpath)
+        for root, dirs, files in os.walk(dirpath, topdown=True):
+            # del dirs[:] # delete the list of dirs, cancelling the recursion
+            # files is the array of files in the tree
+            if not self.recursive_search:
+                dirs.clear()
+            wallpapers.extend(
+                [os.path.join(root, x) for x in files
+                 if self.__should_add_file(wallpapers, x)])
```

Comment:
Is this a bug? We're extending the wallpapers... and not returning anything... or am I not understanding something here?

Reply:
What do you mean by returning? For the method' scope or the call to `extend`?
If you talk about `extend`, this is not a bug, `extend` returns `None` and modifies the variable `wallpapers`. Unless there is something I am not aware of (maybe a change between python2 and 3?). It works on my machine, using python3.
If you talk about the whole function, I just chose not to return anything.

Reply:
Okay. I think I understand what you did there. You're modifying `wallpapers` via the method `__find_wallpapers` instead of `__find_wallpapers_in_dir` hence no return. I got it now.

As for changes between py2/py3, I'm not seeing any issues. IIRC `list()` instead of `[]` might lead to some unicode bugs related to filenames so it could be something to sneeze about. Not sure.

Still digging. Thanks for the help. 👍